src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/TupleTypeDecoder.vb

```diff
@@ -0,0 +1,308 @@
+
+                Case SymbolKind.ArrayType
+                    Return DecodeArrayType(DirectCast(type, ArrayTypeSymbol))
+
+                Case Else
+                    Throw ExceptionUtilities.UnexpectedValue(type.TypeKind)
+            End Select
+        End Function
+
+        Private Function DecodeNamedType(type As NamedTypeSymbol) As NamedTypeSymbol
+            ' First decode the type arguments
+            Dim typeArgs = type.TypeArgumentsNoUseSiteDiagnostics
+            Dim decodedArgs = DecodeTypeArguments(typeArgs)
+
+            Dim decodedType = type
+
+            ' Now check the container
+            Dim containingType = type.ContainingType
+            Dim decodedContainingType As NamedTypeSymbol = Nothing
+
+            If containingType IsNot Nothing AndAlso containingType.IsGenericType Then
+                decodedContainingType = DecodeNamedType(containingType)
+                Debug.Assert(decodedContainingType.IsGenericType)
+            Else
+                decodedContainingType = containingType
+            End If
+
+            ' Replace the type if necessary
+            Dim containerChanged = decodedContainingType IsNot containingType
+            Dim typeArgsChanged = typeArgs <> decodedArgs
+            If typeArgsChanged OrElse containerChanged Then
+                Dim newTypeArgs = If(type.HasTypeArgumentsCustomModifiers,
+                                     decodedArgs.ZipAsArray(type.TypeArgumentsCustomModifiers, Function(t, m) New TypeWithModifiers(t, m)),
+                                     decodedArgs.SelectAsArray(Function(t) New TypeWithModifiers(t, Nothing)))
+
+                If (containerChanged) Then
+                    decodedType = decodedType.OriginalDefinition.AsMember(decodedContainingType)
+                    ' If the type is nested, e.g. Outer(of T).Inner(of V), then Inner is definitely
+                    ' not a tuple, since we know all tuple-compatible types (System.ValueTuple)
+                    ' are not nested types. Thus, it is safe to return without checking if
+                    ' Inner is a tuple.
+                    Return If(type.TypeParameters.IsEmpty,
+                                type,
+                                Construct(type, newTypeArgs))
+                End If
+
+                decodedType = Construct(type, newTypeArgs)
+            End If
+
+            ' Now decode into a tuple, if it is one
+            Dim tupleCardinality As Integer
+            If decodedType.IsTupleCompatible(tupleCardinality) Then
+                Dim elementNames = EatElementNamesIfAvailable(tupleCardinality)
+
+                Debug.Assert(elementNames.IsDefault OrElse elementNames.Length = tupleCardinality)
+
+                decodedType = If(elementNames.IsDefault,
+                                TupleTypeSymbol.Create(decodedType),
+                                TupleTypeSymbol.Create(decodedType, elementNames))
```

Comment:
`If` is unnecessary: `decodedType = TupleTypeSymbol.Create(decodedType, elementNames)`